### PR TITLE
Fix multi jar publishing

### PR DESCRIPTION
**Pull request by alpar-t**
Versions of Gradle before 4.9 don't support project that depend on other
projects which have multiple jar publications, see gradle/gradle#1061.

It just so happens that we upgraded master and 6.x to 4.9 recently, but
6.3 is still on Gradle 4.5 so the multi publications will break there.

This change removes the client jar publication, since now the artifactId
is the only one that differes, and adds the nebula configuration to the
plugin so we don't rely on it's presence in the build scipt to generate
a correct pom. Without the plugin applied, the generated dependencies
would be empty.

Relates elastic/elasticsearch#31946

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy [hunk 1]
```diff
@@ -169,21 +165,42 @@ public class PluginBuildPlugin extends BuildPlugin {
...
-            Files.copy(
-                    project.tasks.generatePomFileForClientJarPublication.outputs.files.singleFile.toPath(),
-                    jarFile.resolveSibling(clientPomFileName),
-                    StandardCopyOption.REPLACE_EXISTING
-            )
+            Node pom = new XmlParser()
+                    .parse(project.tasks.generatePomFileForNebulaPublication.outputs.files.singleFile)
+            pom.artifactId[0].value = project.name + "-client"
+            jarFile.resolveSibling(clientPomFileName).toFile().text = XmlUtil.serialize(pom)
+            // Groovy has an odd way of formatting the XML, fix it up3
```
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: extra `3`?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy [hunk 2]
```diff
@@ -169,21 +165,42 @@ public class PluginBuildPlugin extends BuildPlugin {
...
+    static final Pattern GIT_PATTERN = Pattern.compile(/git@([^:]+):([^\.]+)\.git/)
+
     /** Adds a task to move jar and associated files to a "-client" name. */
     protected static void addClientJarTask(Project project) {
         Task clientJar = project.tasks.create('clientJar')
-        clientJar.dependsOn(project.jar, project.tasks.generatePomFileForClientJarPublication, project.javadocJar, project.sourcesJar)
+        clientJar.dependsOn(project.jar, project.tasks.generatePomFileForNebulaPublication, project.javadocJar, project.sourcesJar)
         clientJar.doFirst {
             Path jarFile = project.jar.outputs.files.singleFile.toPath()
             String clientFileName = jarFile.fileName.toString().replace(project.version, "client-${project.version}")
```
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: It'd be nice if this were declare an output. I know it didn't have it before though.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy [hunk 3]
```diff
@@ -169,21 +165,42 @@ public class PluginBuildPlugin extends BuildPlugin {
         project.artifacts.add('zip', bundle)
     }
 
+    /** Find the reponame. */
+    static String urlFromOrigin(String origin) {
```
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: Is this still used?

**Review by nik9000 — COMMENTED** [review]
The generated POM looks good to me. I left some minor questions about the gradle changes.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy (hunk 3 above)
Comment (atorok) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: yes, by a build script at [distribution/archives/build.gradle](https://github.com/elastic/elasticsearch/blob/master/distribution/archives/build.gradle) we could probably have that use nebula and clean it up, but didn't want to get into that now.

File: modules/transport-netty4/build.gradle [hunk 4]
```diff
@@ -23,6 +23,7 @@
    * fix the hack in the build framework that copies transport-netty4 into the integ test cluster
    * maybe figure out a way to run all tests from core with netty4/network?
  */
+
```
[the remark this answers is not in the document]
Comment (atorok) [reply modules/transport-netty4/build.gradle]: Git annotate says  these were added by @jasontedor, I'm happy to help with it.

**Review by rjernst — COMMENTED** [review]
I'm not sure we actually need this, given there is a workaround (only have client jar or normal jar+pom). Given we don't currently have a need for both, is there any reason not to just apply the same tweak (add back the else condition for configuring the jar pom)?

**Comment by atorok** [comment]
@rjernst I'm ok with moving jar generation so it doesn't happen at the same time with client jar.  
If we don't it would make sense for me to depend on main jar and not client jar directly. Not sure how disruptive that would be. A client jar makes sense to me as a separate Gradle project that really only has the client part and is depended on from the server, but admittedly I am missing a log of context here.

An argument for keeping it might be for plugin authors that build multiple inter dependent plugins, but that's probably an overstretch, as those users can easily do their own pom generation. 

I do think we need to keep from this PR the part about applying nebula, as it is it's too easy to forget that and don't see a reason to rely on the build script to apply it rather than doing it in the plugin. 
The pom without nebula base applied does not have dependencies, so I don't think it's useful in any way.
We could also do it without nebula (as done by client jar prior to this change) or include only base publishing from nebula and leave it up to the build script to decide if it wants scm info as well.

**Comment by rjernst** [comment]
> the part about applying nebula, as it is it's too easy to forget that and don't see a reason to rely on the build script to apply it rather than doing it in the plugin

Publishing is a choice. Plugin authors could be publishing their plugins in many ways. This let's them choose if they want to use nebula. For all the stuff that build-tools pulls in for plugin development, all of it is optional (the tasks can be disabled). AFAIK, publishing is not something that can simply be disabled, because the tasks are generated late and not available when the plugin is applied.

**Comment by rjernst** [comment]
I suspect the fact we always pull in publishing and generate a pom is leftover from the very first iteration of gradle, when we were still always publishing plugin zips to maven. Since we don't do this anymore (and haven't since 5.0 beta), we should change this to only create a pom if configured.

**Comment by atorok** [comment]
Closing in favor of #32351

**Pull request closed without merging**